misk-hibernate/src/main/kotlin/misk/hibernate/SecretColumnType.kt

```diff
@@ -90,4 +93,59 @@ internal class SecretColumnType : UserType, ParameterizedType, TypeConfiguration
   override fun isMutable() = false
 
   override fun sqlTypes() = intArrayOf(Types.VARBINARY)
-}
\ No newline at end of file
+}
+
+internal interface EncryptionAdapter {
+  fun encrypt(plaintext: ByteArray, associatedData: ByteArray?): ByteArray
+  fun decrypt(ciphertext: ByteArray, associatedData: ByteArray?): ByteArray
+}
+
+internal class AeadAdapter(typeConfig: TypeConfiguration, keyName: String) : EncryptionAdapter {
+
+  val aead: Aead
+
+  init {
+    val keyManager = typeConfig.metadataBuildingContext.bootstrapContext.serviceRegistry.injector
+            .getInstance(AeadKeyManager::class.java)
+    try {
+      aead = keyManager[keyName]
+    } catch (ex: KeyNotFoundException) {
+      throw HibernateException("Cannot set field, key $keyName not found")
+    }
+  }
+
+  override fun encrypt(plaintext: ByteArray, associatedData: ByteArray?) : ByteArray {
+    return aead.encrypt(plaintext, associatedData)
+  }
+
+  override fun decrypt(ciphertext: ByteArray, associatedData: ByteArray?) : ByteArray {
+    return aead.decrypt(ciphertext, associatedData)
+  }
+}
+
+internal class DeterministicAeadAdapter(typeConfig: TypeConfiguration, keyName: String)
+  : EncryptionAdapter {
+
+  val daead: DeterministicAead
+
+  init {
+    val keyManager = typeConfig.metadataBuildingContext.bootstrapContext.serviceRegistry.injector
+            .getInstance(DeterministicAeadKeyManager::class.java)
+    try {
+      daead = keyManager[keyName]
+    } catch (ex: KeyNotFoundException) {
+      throw HibernateException("Cannot set field, key $keyName not found")
+    }
+  }
+
+  // DeterministicAEAD throws if associatedData is null, so we pass an empty array if it is.
```

Comment:
This comment should be inside the function body

Reply:
It was initially but it covers both encrypt and decrypt. I can move it in though.